Add the release note for 1.29.0 (#5723)
line/armeria: site/src/pages/release-notes/1.29.0.mdx

- <type://ArmeriaServerConfigurator> now properly overrides the properties set by <type://ArmeriaSettings>. #5009 #5692
- Armeria server now returns a 408 status if a service didn't receive the request fully and
  the request times out. #5579 #5680
- You can now specify a field for gRPC transcoding with `p{\d}` pattern. #5679
- A `NullPointerException` isn't raised anymore when running Armeria with a self-signed certificate. #5669
  It happened when the following conditions are all met:
  - DEBUG-level logging of JDK security event is enabled.
  - A user has no Bouncy Castle dependency.
- A `DecompressionException` isn't ignored anymore in a <type://StreamDecoder>. #5177 #5620
- <type://@Description> annotations at the super class or interface are now used for the parameter description. #5195 #5562
- The specfied `ObjectMapper` to <type://WebClientRequestPreparation#asJson(Class,ObjectMapper)> isn't
  ignored anymore. #5454 #5512
- <type://GrpcExceptionHandlerFunction> is now applied to convert a gRPC cancellations. #5329 #5398

## 🏚️ Deprecations

- <type://PrometheusMeterRegistries> and <type://PrometheusExpositionService> are deprecated. #5698

>   - Use the same classes in the `armeria-promehteus1` module.

Reviewer: ```suggestion
  - Use the same classes in the `armeria-prometheus1` module.
```

- <type://HttpResult> is deprecated. Use <type://ResponseEntity> instead. #4910 #5586
- <type://GraphqlConfigurator> and various setter methods for building a `GraphQL` instance are deprecated.
  - Use <type://GraphqlServiceBuilder#graphql(GraphQL)> instead. #5269

## ☢️ Breaking changes

- Micrometer 1.13.0 uses Prometheus Java client 1.x. #5698
  - If you want to keep the old behavior that uses Prometheus Java client 0.x,
    use the `io.micrometer:micrometer-registry-prometheus-simpleclient:1.13.0` module.

>   - If you want to use Prometheus Java client 1.x, add `com.linecorp.armeria:armeria-promehteus1` module.

Reviewer: ```suggestion
  - If you want to use Prometheus Java client 1.x, add `com.linecorp.armeria:armeria-prometheus1` module.
```

  ```protobuf
  rpc Watch(WatchRequest) returns (WatchResponse) {
    option (google.api.http) = {
      post: "/v1:watch" // 👈👈👈
      body: "*"
    };
  }
  ```
- You can now specify when to remove multipart temporary files using <type://MultipartRemovalStrategy>. #5652 #5653
  ```java
  Server
    .builder()
    .multipartRemovalStrategy(MultipartRemovalStrategy.ON_RESPONSE_COMPLETION)
    ...
  ```
- A <type://ClientConnectionTimings> now includes the time spent on the TLS handshake. #3647 #5647
  ```java
  ClientConnectionTimings timings = ...
  timings.tlsHandshakeDurationNanos();
  ```
- You can now configure <type://TlsEngineType> using <type://ClientFactoryBuilder> or <type://ServerBuilder>. #4962
  ```java
  ClientFactory
    .builder()
    .tlsEngineType(TlsEngineType.OPENSSL)
    .build();
  ```
- You can now easily find both dynamic and static decorators that handle a request
  via <type://ServiceRequestContext.findService(Class)>. #5670
  ```java
  ServerBuilder sb = ...
  sb.service("/foo", fooService);
  ...
  ServiceRequestContext ctx = ...
  assert ctx.findService(FooService.class) == fooService;

>   ```

Reviewer 2: I feel like this is not the best example. Can you use an example that uses a route decorator? `CorsService` could be looked for.

- You can now use the marshaller specified by gRPC `MethodDescriptor`. #5103 #5630
  ```java
  GrpcClientBuilder builder = ...
  builder.useMethodMarshaller(true);
  ```
- You can now programmatically retrieve the server side metrics via <type://ServerMetrics>. #4992 #5627
  ```java
  ServerMetrics metrics = serverConfig.serverMetrics();
  metrics.activeConnections();
  metrics.pendingRequests();
  metrics.activeRequests();
  ```
- You can now use a <type://CoroutineHttpService> that runs your service in a coroutine scope. #5442 #5603
  ```kotlin
  ServerBuilder sb = ...
  sb.service(
    "/hello",
    CoroutineHttpService { ctx, req ->
      HttpResponse.of("hello world")
    })
  ```
- You can now specify options for an <type://HttpService> via <type://ServiceOptions>. #5071 #5574
  ```java
  HttpService httpService = new HttpService() {
    ...
    @Override
    public ServiceOptions options() {

>       return customizedOptions;

Reviewer 2: Please make this example more useful. How about defining a static final field of `ServiceOptions`?

    }
  };

  // Or use annotation for an annotated service.
  class MyService {
      @ServiceOption(requestTimeoutMillis = 10000, maxRequestLength = 1024)
      @Get("/hello")
      public HttpResponse hello() {
          return HttpResponse.of("Hello!");
      }
  }
  ```
- You can now inject a custom attribute from a <type://ServiceRequestContext> to an annotated service
  using <type://@Attribute> annotation. #5514 #5547
  ```java
  class MyAttributes {
    public static final AttributeKey<String> USERNAME =
      AttributeKey.valueOf(MyAttributes.class, "USERNAME");
  }

  class MyAnnotatedService {

    @Get("/hello")
    public HttpResponse hello(

>       @Attribute(prefix = MyAttributes.class, value = "USERNAME") // 👈👈👈
>         String username) { ... }

Reviewer 2: ```suggestion
      @Attribute(prefix = MyAttributes.class, value = "USERNAME") // 👈👈👈
      String username) { ... }
```

  }
  ```
- You can now specify a graceful shutdown timeout for a <type://ClientFactory>. #5470 #5489
  ```java
  ClientFactory
    .builder()
    .http2GracefulShutdownTimeoutMillis(1000)
    .build();
  ```

> - You can now specify an <type://EndpointGroup> to build a <type://ClientRequestContext>. #5292 #5298

Reviewer 2: ```suggestion
- You can now specify an <type://EndpointGroup> when building a <type://ClientRequestContext>. #5292 #5298
```

- A `NullPointerException` isn't raised anymore when running Armeria with a self-signed certificate. #5669
  It happened when the following conditions are all met:
  - DEBUG-level logging of JDK security event is enabled.
  - A user has no Bouncy Castle dependency.
- A `DecompressionException` isn't ignored anymore in a <type://StreamDecoder>. #5177 #5620
- <type://@Description> annotations at the super class or interface are now used for the parameter description. #5195 #5562
- The specfied `ObjectMapper` to <type://WebClientRequestPreparation#asJson(Class,ObjectMapper)> isn't
  ignored anymore. #5454 #5512
- <type://GrpcExceptionHandlerFunction> is now applied to convert a gRPC cancellations. #5329 #5398

## 🏚️ Deprecations

- <type://PrometheusMeterRegistries> and <type://PrometheusExpositionService> are deprecated. #5698
  - Use the same classes in the `armeria-prometheus1` module.
- <type://HttpResult> is deprecated. Use <type://ResponseEntity> instead. #4910 #5586
- <type://GraphqlConfigurator> and various setter methods for building a `GraphQL` instance are deprecated.
  - Use <type://GraphqlServiceBuilder#graphql(GraphQL)> instead. #5269

## ☢️ Breaking changes

> - Micrometer 1.13.0 uses Prometheus Java client 1.x. #5698

Reviewer 2: I think this sentence is somewhat misleading. Micrometer is agnostic to metric collector implementations and it supports both Prometheus Java client versions. How about rephrasing?

```suggestion
- We updated Micrometer to 1.13.0 that has breaking changes in its Prometheus support. #5698
```

  can now access the current <type://ServiceRequestContext> if available. #5634
- <type://SamlSingleSignOnHandler#loginFailed(ServiceRequestContext,AggregatedHttpRequest,MessageContext,Throwable)>
  can now access the <type://MessageContext> for error handling if available. #5401 #5622
- The copied cURL command from a doc service, is now quoted correctly. #5566

## 🛠️ Bug fixes

- Fixed a bug where <type://RetryingClient> gets to deadlock when <type://OAuth2Client> fails to obtain
  an access token. #5715
- Armeria client always uses HTTP/2 connection preface for `h2c`, regardless of
  the value of <type://ClientFactoryOptions#useHttp2Preface()>. #5706
- <type://ArmeriaServerConfigurator> now properly overrides the properties set by <type://ArmeriaSettings>. #5009 #5692
- Armeria server now returns a 408 status if a service didn't receive the request fully and
  the request times out. #5579 #5680
- A `NullPointerException` isn't raised anymore when running Armeria with a self-signed certificate. #5669
  It happened when the following conditions are all met:
  - DEBUG-level logging of JDK security event is enabled.
  - A user has no Bouncy Castle dependency.

> - A `DecompressionException` isn't ignored anymore in a <type://StreamDecoder>. #5177 #5620

Reviewer 3: It was rescheduled to 1.30.0
```suggestion
```

Author: Thanks fixed. 😉

---
date: 2024-06-05
---

## 🌟 New features

- Armeria [HTTP/JSON to GRPC transcoding](https://cloud.google.com/endpoints/docs/grpc/transcoding) service now
  supports [Custom methods](https://cloud.google.com/apis/design/custom_methods). #5613
  ```protobuf
  rpc Watch(WatchRequest) returns (WatchResponse) {
    option (google.api.http) = {
      post: "/v1:watch" // 👈👈👈
      body: "*"
    };
  }
  ```

> - You can now specify when to remove multipart temporary files using <type://MultipartRemovalStrategy>. #5652 #5653

Reviewer 4: Optional) It may be worth noting that the default has changed to `ON_RESPONSE_COMPLETION` as well.

  ```java
  static final ServiceOptions SERVICE_OPTIONS =
    ServiceOptions
      .builder()
      .requestTimeoutMillis(5000)
      .maxRequestLength(1024)
      .requestAutoAbortDelayMillis(1000)
      .build();

  HttpService httpService = new HttpService() {
    ...
    @Override
    public ServiceOptions options() {
      return SERVICE_OPTIONS;
    }
  };

  // Or use annotation for an annotated service.
  class MyService {
      @ServiceOption(requestTimeoutMillis = 5000, maxRequestLength = 1024)
      @Get("/hello")
      public HttpResponse hello() {
          return HttpResponse.of("Hello!");
      }
  }
  ```
- You can now inject a custom attribute from a <type://ServiceRequestContext> to an annotated service
  using <type://@Attribute> annotation. #5514 #5547
  ```java
  class MyAttributes {
    public static final AttributeKey<String> USERNAME =
      AttributeKey.valueOf(MyAttributes.class, "USERNAME");
  }

  class MyAnnotatedService {

    @Get("/hello")
    public HttpResponse hello(
      @Attribute(prefix = MyAttributes.class, value = "USERNAME") // 👈👈👈
      String username) { ... }
  }
  ```

> - You can now specify a graceful shutdown timeout for a <type://ClientFactory>. #5470 #5489

Reviewer 3: Technically, it is a graceful timeout for HTTP/2 connection but not `ClientFactory`.
```suggestion
- You can now specify a graceful shutdown timeout for an HTTP/2 connection 
  in <type://ClientFactory>. #5470 #5489
```

  ```java
  ClientFactory
    .builder()
    .http2GracefulShutdownTimeoutMillis(1000)
    .build();
  ```
- You can now specify an <type://EndpointGroup> when building a <type://ClientRequestContext>. #5292 #5298
  ```java
  ClientRequestContext ctx =
    ClientRequestContext
      .builder(request)
      .endpointGroup(endpointGroup)
      .build();
  ```
- You can now set a `GraphQL` instance directly to the <type://GraphqlServiceBuilder>. #5269
  ```java
  GraphQL graphQL = new GraphQL.Builder(graphqlSchema).build();
  GraphqlServiceBuilder builder = ...
  builder.graphql(graphQL);
- You can now specify a delay to close an HTTP/1.1 connection from the server side, allowing the client an
  opportunity for active closing. #4849 #5616

## 📈 Improvements

> - An <type://XdsEndpointGroup> now supports [Slow start mode](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/load_balancing/slow_start). #5688 #5693

Reviewer 3: nit: Drop `An` for simplicity?
```suggestion
- <type://XdsEndpointGroup> now supports [Slow start mode](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/load_balancing/slow_start). #5688 #5693
```

- You can now view connection related information in the request logs. #5730
- CORS headers are added to responses even if an exception is raised while handling a request. #5632
- <type://AnnotatedService> is now available through a public API. #5382 #5628
- The metrics for failed requests while retrying now include the cause of the failure. #5405 #5583
- The <type://RequestLog> from a retrying client now includes the current attempt number. #5716 #5719
- A colon is now allowed in the path for binding a service. #4577 #5676
- Timer histogram metrics now repects user-provided `serviceLevelObjectives`. #5661
- <type://ServiceErrorHandler#renderStatus(ServiceRequestContext,RequestHeaders,HttpStatus,String,Throwable)>
  can now access the current <type://ServiceRequestContext> if available. #5634
- <type://SamlSingleSignOnHandler#loginFailed(ServiceRequestContext,AggregatedHttpRequest,MessageContext,Throwable)>
  can now access the <type://MessageContext> for error handling if available. #5401 #5622
- The copied cURL command from a doc service, is now quoted correctly. #5566

## 🛠️ Bug fixes

> - <type://HealthCheckedEndpointGroup> no longer raises an <type://EndpointSelectionTimeoutException> when
>   a server is stopped by the JVM shutdown hook. #5718

Reviewer 3: #5718 is removed from 1.29.0
```suggestion
```

  ignored anymore. #5454 #5512
- <type://GrpcExceptionHandlerFunction> is now applied to convert a gRPC cancellations. #5329 #5398
- In Spring integration, the default <type://ClientFactory> is now gracefully closed after the server shuts down. #5742
- A connection is not reused anymore after a <type://WriteTimeoutException> is raised. #5738

## 🏚️ Deprecations

- <type://PrometheusMeterRegistries> and <type://PrometheusExpositionService> are deprecated. #5698
  - Use the same classes in the `armeria-prometheus1` module.
- <type://HttpResult> is deprecated. Use <type://ResponseEntity> instead. #4910 #5586
- <type://GraphqlConfigurator> and various setter methods for building a `GraphQL` instance are deprecated.
  - Use <type://GraphqlServiceBuilder#graphql(GraphQL)> instead. #5269

## ☢️ Breaking changes

- We updated Micrometer to 1.13.0 that has breaking changes in its Prometheus support. #5698
  - If you want to keep the old behavior that uses Prometheus Java client 0.x,
    use the `io.micrometer:micrometer-registry-prometheus-simpleclient:1.13.0` module.
  - If you want to use Prometheus Java client 1.x, add `com.linecorp.armeria:armeria-prometheus1` module.

>   - More details can be found in the [migration guide](https://github.com/micrometer-metrics/micrometer/wiki/1.13-Migration-Guide).

Reviewer 3: nit: Because it isn't our migration guide.
```suggestion
  - More details can be found in the [Micrometer migration guide](https://github.com/micrometer-metrics/micrometer/wiki/1.13-Migration-Guide).
```

> - The following builder classes are now have the `SELF` type parameter. #5733

Reviewer 4: ```suggestion
- The following builder classes now have the `SELF` type parameter. #5733
```

  - <type://AbstractDnsResolverBuilder>
  - <type://AbstractRuleBuilder>
  - <type://AbstractRuleWithContentBuilder>
  - <type://AbstractCircuitBreakerMappingBuilder>
  - <type://AbstractDynamicEndpointGroupBuilder>
  - <type://DynamicEndpointGroupSetters>
  - <type://AbstractCuratorFrameworkBuilder>

## ⛓ Dependencies

- Blockhound 1.0.8.RELEASE → 1.0.9.RELEASE
- Control Plane 1.0.44 → 1.0.45
- GraphQL Kotlin 7.0.2 → 7.1.1
- gRPC Java 1.63.0 → 1.64.0

> - Guava 33.1.0-jre → 33.2.0-jre

Reviewer 3: It is not a transitive dependency.  
```suggestion
```

---

> date: 2024-06-11

Author: ```suggestion
date: 2024-06-12
```
